modules/uid2IdSystem_shared.js

```diff
@@ -160,18 +173,508 @@ function refreshTokenAndStore(baseUrl, token, clientId, storageManager, _logInfo
+        if (!this.isStoredTokenFromSameIdentity(storedTokens, identity)) {
+          _logInfo(
+            'CSTG supplied new identity - ignoring stored value.',
+            storedTokens.originalIdentity,
+            cstgIdentity
+          );
+          // Stored token wasn't originally sourced from the provided identity - ignore the stored value. A new user has logged in?
+          return true;
+        }
+      }
+      return false;
+    },
+
+    async generateTokenAndStore(
+      baseUrl,
+      cstgOpts,
+      cstgIdentity,
+      storageManager,
+      _logInfo,
+      _logWarn
+    ) {
+      _logInfo('UID2 cstg opts provided: ', JSON.stringify(cstgOpts));
+      const client = new UID2CstgApiClient(
+        { baseUrl, cstg: cstgOpts },
+        _logInfo,
+        _logWarn
+      );
+      const response = await client.generateToken(cstgIdentity);
+      _logInfo('CSTG endpoint responded with:', response);
+      const tokens = {
+        originalIdentity: this.encodeOriginalIdentity(cstgIdentity),
+        latestToken: response.identity,
+      };
+      storageManager.storeValue(tokens);
+      return tokens;
+    },
+
+    isStoredTokenFromSameIdentity(storedTokens, identity) {
+      if (!storedTokens.originalIdentity) return false;
+      return (
+        cyrb53Hash(identity, storedTokens.originalIdentity.salt) ===
+        storedTokens.originalIdentity.identity
+      );
+    },
+
+    encodeOriginalIdentity(identity) {
+      const identityValue = Object.values(identity)[0];
+      const salt = Math.floor(Math.random() * Math.pow(2, 32));
+      return {
+        identity: cyrb53Hash(identityValue, salt),
+        salt,
+      };
+    },
+  };
+
+  class UID2DiiNormalization {
+    static EMAIL_EXTENSION_SYMBOL = '+';
+    static EMAIL_DOT = '.';
+    static GMAIL_DOMAIN = 'gmail.com';
+
+    static isBase64Hash(value) {
+      if (!(value && value.length === 44)) {
+        return false;
+      }
+
+      try {
+        return btoa(atob(value)) === value;
+      } catch (err) {
+        return false;
+      }
+    }
+
+    static isNormalizedPhone(phone) {
+      return /^\+[0-9]{10,15}$/.test(phone);
+    }
+
+    static normalizeEmail(email) {
+      if (!email || !email.length) return;
+
+      const parsedEmail = email.trim().toLowerCase();
+      if (parsedEmail.indexOf(' ') > 0) return;
+
+      const emailParts = this.splitEmailIntoAddressAndDomain(parsedEmail);
+      if (!emailParts) return;
+
+      const { address, domain } = emailParts;
+
+      const emailIsGmail = this.isGmail(domain);
+      const parsedAddress = this.normalizeAddressPart(
+        address,
+        emailIsGmail,
+        emailIsGmail
+      );
+
+      return parsedAddress ? `${parsedAddress}@${domain}` : undefined;
+    }
+
+    static splitEmailIntoAddressAndDomain(email) {
+      const parts = email.split('@');
+      if (
+        !parts.length ||
+        parts.length !== 2 ||
```

Comment:
first clause, `!parts.length`, is redundant since if it's falsy it's also always `!== 2`

Reply:
Good point, I removed that, thanks